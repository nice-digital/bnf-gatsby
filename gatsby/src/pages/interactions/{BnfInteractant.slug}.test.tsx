import {
	render,
	waitFor,
	screen,
	within,
	fireEvent,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import InteractantPage, {
	query,
	type InteractantPageProps,
} from "./{BnfInteractant.slug}";

const interactant: InteractantPageProps["data"]["bnfInteractant"] = {
	// Note deliberate use of HTML within the title for testing stripping of tags
	title: "Anti-D (Rh<sub>0</sub>) immunoglobulin",
	drug: {
		title: "acarbose",
		slug: "acarbose",
	},
	interactions: [
		{
			interactant: {
				title: "Pancreatin",
				drug: {
					slug: "pancreatin",
				},
			},
			messages: [
				{
					additiveEffect: false,
					evidence: "Theoretical",
					message:
						'<span class="substance-primary" data-monograph-sid="_743802912" data-sid="_743802912">pancreatin</span> <span class="effectQualifier">is predicted to</span> <span class="effect">decrease</span> <span class="parameter">the effects of</span> <span class="substance-primary" data-monograph-sid="_426793441" data-sid="_426793441">acarbose</span>. <span class="action">Manufacturer advises avoid</span>.  ',
					severity: "Moderate",
					severityOrder: 3,
				},
			],
		},
		{
			interactant: {
				title: "Canagliflozin test",
				drug: {
					slug: "canagliflozin",
				},
			},
			messages: [
				{
					additiveEffect: true,
					evidence: null,
					message:
						'<p>Both <span class="substance-primary" data-monograph-sid="_426793441" data-sid="_426793441">acarbose</span> and <span class="substance-primary" data-monograph-sid="_629412592" data-sid="_629412592">canagliflozin</span> can increase the risk of hypoglycaemia.</p>',
					severity: "Normal",
					severityOrder: 1,
				},
			],
		},
		{
			interactant: {
				title: "Test interactant",
				drug: {
					slug: "testinteractant",
				},
			},
			messages: [
				{
					additiveEffect: true,
					evidence: null,
					message: "Test message 1",
					severity: "Normal",
					severityOrder: 1,
				},
				{
					additiveEffect: true,
					evidence: null,
					message: "Test message 2 (severe)",
					severity: "Severe",
					severityOrder: 4,
				},
			],
		},
	],
	supplementaryInformation: [],
};

const dataProp: InteractantPageProps["data"] = {
	bnfInteractant: interactant,
};

describe("InteractantPage", () => {
	it("should match snapshot for graphql query", () => {
		expect(query).toMatchSnapshot();
	});

	describe("SEO", () => {
		it("should strip HTML tags for page title", async () => {
			render(<InteractantPage data={dataProp} />);

			await waitFor(() => {
				expect(document.title).toStartWith(
					"Anti-D (Rh0) immunoglobulin | Interactions |"
				);
			});
		});

		it("should strip HTML tags for templated meta description", async () => {
			render(<InteractantPage data={dataProp} />);

			await waitFor(() => {
				expect(
					document
						// eslint-disable-next-line testing-library/no-node-access
						.querySelector("meta[name='description']")
				).toHaveAttribute(
					"content",
					"See the list of drugs that interact with Anti-D (Rh0) immunoglobulin. Includes information on severity of interaction and the level of evidence for it."
				);
			});
		});
	});

	describe("Page header", () => {
		it("should add content start skip link target id to page header", () => {
			render(<InteractantPage data={dataProp} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			// eslint-disable-next-line testing-library/no-node-access
			expect(heading1.parentElement).toHaveAttribute("id", "content-start");
		});

		it("should render heading 1 with current page title", () => {
			render(<InteractantPage data={dataProp} />);
			const heading1 = screen.getByRole("heading", {
				level: 1,
			});
			expect(heading1).toHaveTextContent(
				"Anti-D (Rh0) immunoglobulin Interactions"
			);
		});
	});

	describe("Breadcrumbs", () => {
		it.each([
			["NICE", "https://www.nice.org.uk/"],
			["BNF", "/"],
			["Interactions", "/interactions/"],
		])(
			"should render default '(%s)' breadcrumb",
			(breadcrumbText, expectedHref) => {
				render(<InteractantPage data={dataProp} />);

				const breadcrumbNav = screen.getByRole("navigation", {
					name: "Breadcrumbs",
				});
				const breadcrumb = within(breadcrumbNav).getByRole("link", {
					name: breadcrumbText,
				});

				expect(breadcrumb).toHaveAttribute("href", expectedHref);
			}
		);

		it("should render current page breadcrumb without link", () => {
			render(<InteractantPage data={dataProp} />);

			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const currentPageCrumb = within(breadcrumbNav).getByText(
				"Anti-D (Rh0) immunoglobulin"
			);
			expect(currentPageCrumb).toBeInTheDocument();
			expect(currentPageCrumb.tagName).toBe("SPAN");
		});
	});

	describe("body", () => {
		it("should render 'has no specific interactions information' when there are no interactions", () => {
			render(
				<InteractantPage
					data={{ bnfInteractant: { ...interactant, interactions: [] } }}
				/>
			);
			expect(
				screen.getByText(
					(_content, node) =>
						node?.textContent ===
						"Anti-D (Rh0) immunoglobulin has no specific interactions information."
				)
			).toBeInTheDocument();
		});

		it("should match snapshot for page body", () => {
			render(<InteractantPage data={dataProp} />);
			expect(screen.getByRole("main")).toMatchSnapshot();
		});

		it("should link to drug monograph page for matching drug", () => {
			render(<InteractantPage data={dataProp} />);
			const monographLink = screen.getByRole("link", {
				name: `View ${interactant.drug?.title} monograph page`,
			});
			expect(monographLink).toHaveAttribute(
				"href",
				`/drugs/${interactant.drug?.slug}/`
			);
		});

		it("should render a list of interactions, one per interactant", () => {
			render(<InteractantPage data={dataProp} />);
			const interactionList = screen.getByRole("list", {
				name: `List of interactions for ${interactant.drug?.title}`,
			});
			// Have to disable linting on no-node-access, as nested list items mean that getAllByRole returns too many elements
			// eslint-disable-next-line testing-library/no-node-access
			expect(interactionList.children).toHaveLength(
				interactant.interactions.length
			);
		});
	});

	describe("sorting", () => {
		beforeEach(() => {
			window.dataLayer = [];
		});

		it("should toggle between the two different sorting buttons whenever one is pressed", () => {
			render(<InteractantPage data={dataProp} />);

			const nameSortButtonArgs: [string, { name: string }] = [
				"button",
				{
					name: "Sort by: Name",
				},
			];
			const severitySortButtonArgs: [string, { name: string }] = [
				"button",
				{
					name: "Sort by: Severity",
				},
			];

			// Have to use queryByRole instead of getByRole for nullable queries (i.e. queries that may return no elements)
			// This is because we want to check that certain elements *don't* exist and have been removed from the DOM
			let nameSortButton: HTMLElement | null = screen.queryByRole(
				...nameSortButtonArgs
			);
			let severitySortButton: HTMLElement | null = screen.getByRole(
				...severitySortButtonArgs
			);
			expect(nameSortButton).not.toBeInTheDocument();
			expect(severitySortButton).toBeInTheDocument();

			userEvent.click(severitySortButton);
			nameSortButton = screen.getByRole(...nameSortButtonArgs);
			severitySortButton = screen.queryByRole(...severitySortButtonArgs);
			expect(nameSortButton).toBeInTheDocument();
			expect(severitySortButton).not.toBeInTheDocument();

			userEvent.click(nameSortButton);
			nameSortButton = screen.queryByRole(...nameSortButtonArgs);
			severitySortButton = screen.getByRole(...severitySortButtonArgs);
			expect(nameSortButton).not.toBeInTheDocument();
			expect(severitySortButton).toBeInTheDocument();
		});
		it("should sort the interaction results alphabetically by default", () => {
			render(<InteractantPage data={dataProp} />);
			expect(
				screen
					.getAllByRole("heading", { level: 3 })
					.map((heading) => heading.textContent)
			).toStrictEqual(["Canagliflozin test", "Pancreatin", "Test interactant"]);
		});

		it("should sort the interaction results by severity after hitting the 'Sort by severity' button", () => {
			render(<InteractantPage data={dataProp} />);
			const sortButton = screen.getByRole("button", {
				name: "Sort by: Severity",
			});
			userEvent.click(sortButton);
			expect(
				screen
					.getAllByRole("heading", { level: 3 })
					.map((heading) => heading.textContent)
			).toStrictEqual(["Test interactant", "Pancreatin", "Canagliflozin test"]);
		});

		it("should limit the result set when a filter value is supplied", () => {
			render(<InteractantPage data={dataProp} />);
			const filterButton = screen.getByRole("button", {
				name: "Filter",
			});
			userEvent.type(screen.getByRole("textbox"), "Test");
			userEvent.click(filterButton);
			expect(
				screen
					.getAllByRole("heading", { level: 3 })
					.map((heading) => heading.textContent)
			).toStrictEqual(["Canagliflozin test", "Test interactant"]);
		});

		it("should successfully list a filtered result set in severity order when specified", () => {
			render(<InteractantPage data={dataProp} />);
			const filterButton = screen.getByRole("button", {
				name: "Filter",
			});
			userEvent.type(screen.getByRole("textbox"), "Test");
			userEvent.click(filterButton);

			const sortButton = screen.getByRole("button", {
				name: "Sort by: Severity",
			});
			userEvent.click(sortButton);

			expect(
				screen
					.getAllByRole("heading", { level: 3 })
					.map((heading) => heading.textContent)
			).toStrictEqual(["Test interactant", "Canagliflozin test"]);
		});

		it("should render supplementary information when supplied", () => {
			render(
				<InteractantPage
					data={{
						bnfInteractant: {
							...interactant,
							supplementaryInformation: [
								{
									title: "Test supplementary info 1",
									information:
										"<p>Supplementary info test information element 1</p>",
								},
								{
									title: "Test supplementary info 2",
									information:
										"<p>Supplementary info test information element 2</p>",
								},
							],
						},
					}}
				/>
			);

			expect(
				screen.getByRole("heading", {
					level: 2,
					name: "Test supplementary info 1",
				})
			).toBeInTheDocument();

			expect(
				screen.getByRole("heading", {
					level: 2,
					name: "Test supplementary info 2",
				})
			).toBeInTheDocument();

			expect(
				screen.getByText("Supplementary info test information element 1")
			).toBeInTheDocument();
			expect(
				screen.getByText("Supplementary info test information element 2")
			).toBeInTheDocument();
		});
	});

	describe("filtering", () => {
		it("should push a formSubmit event to the data layer when filter button submits form", async () => {
			window.dataLayer = [];
			render(<InteractantPage data={dataProp} />);

			const inputElement = screen.getByLabelText("Filter by drug name");

			const submitButton = screen.getByRole("button", {
				name: "Filter",
			});

			userEvent.type(inputElement, "some filter text");

			fireEvent.click(submitButton);

			await waitFor(() => {
				expect(window.dataLayer[0]).toStrictEqual({
					event: "formSubmit",
					formText: "some filter text",
				});
			});
		});
	});
});
