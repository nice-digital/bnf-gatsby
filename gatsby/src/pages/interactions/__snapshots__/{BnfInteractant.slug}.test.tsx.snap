// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractantPage body should match snapshot for page body 1`] = `
<main>
  <div
    class="container"
  >
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        class="breadcrumbs"
      >
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="https://www.nice.org.uk/"
          >
            NICE
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/"
          >
            BNF
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/interactions/"
          >
            Interactions
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <span>
            Anti-D (Rh0) immunoglobulin
          </span>
        </li>
      </ol>
    </nav>
    <script
      type="application/ld+json"
    >
      {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/",
        "name": "BNF"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/interactions/",
        "name": "Interactions"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Anti-D (Rh0) immunoglobulin"
      }
    }
  ]
}
    </script>
    <div
      class="page-header"
      id="content-start"
    >
      <h1
        class="page-header__heading"
      >
        <span
          class="page-header__pre-heading"
        >
          <span>
            Anti-D (Rh
            <sub>
              0
            </sub>
            ) immunoglobulin 
          </span>
        </span>
        Interactions
      </h1>
      <p
        class="page-header__lead"
      >
        <a
          href="/drugs/acarbose/"
        >
          View 
          <span>
            acarbose
          </span>
           
          monograph page
        </a>
      </p>
    </div>
    <p
      class="interactionInformation"
    >
      <span>
        Anti-D (Rh
        <sub>
          0
        </sub>
        ) immunoglobulin
      </span>
       has the following interaction information:
    </p>
    <div
      class="grid"
    >
      <div
        class="rightCol"
      >
        <div
          class="informationPanel"
        >
          <h2
            class="informationPanelHeading"
          >
            Drug interaction information
          </h2>
          <p
            class="severeMessage"
          >
            Severe interactions are highlighted with a red marker
          </p>
          <p>
            <a
              href="/interactions/introduction/"
            >
              Find out more about BNF interactions information
            </a>
          </p>
        </div>
      </div>
      <div
        class="leftCol"
      >
        <section
          class="filterPanel hide-print"
        >
          <h2
            class="visually-hidden"
          >
            Filters and sorting
          </h2>
          <form
            class="filterForm"
          >
            <div
              class="input filterInput"
            >
              <label
                class="input__label"
                for="drugNameInput"
              >
                Filter by drug name
              </label>
              <input
                class="input__input"
                id="drugNameInput"
                name="drugNameInput"
                placeholder="Enter drug name"
                type="text"
                value=""
              />
            </div>
            <button
              class="btn btn--secondary filterButton"
              type="button"
            >
              Filter
            </button>
          </form>
          <div
            class="sortControls"
          >
            <strong>
              Sorted by: 
            </strong>
            <span
              class="sortButtonLabel"
            >
              Name | 
            </span>
            <button
              class="sortButton"
            >
              Sort by: Severity
            </button>
          </div>
        </section>
        <div
          class="resultCount"
        >
          Showing 
          3
           of 
          3
        </div>
        <section
          aria-live="polite"
        >
          <h2
            class="visually-hidden"
            id="interactions-list-heading"
          >
            List of interactions for
             
            <span>
              acarbose
            </span>
          </h2>
          <ol
            aria-labelledby="interactions-list-heading"
            class="interactionsList"
          >
            <li
              class="interactionsListItem"
            >
              <h3
                class="interactantTitle"
              >
                <a
                  href="/drugs/canagliflozin/"
                >
                  Canagliflozin test
                </a>
              </h3>
              <ul
                class="messageList"
              >
                <li
                  class="message"
                >
                  <div
                    class=""
                  >
                    <div>
                      <p>
                        Both 
                        <span
                          class="substance-primary"
                          data-monograph-sid="_426793441"
                          data-sid="_426793441"
                        >
                          acarbose
                        </span>
                         and 
                        <span
                          class="substance-primary"
                          data-monograph-sid="_629412592"
                          data-sid="_629412592"
                        >
                          canagliflozin
                        </span>
                         can increase the risk of hypoglycaemia.
                      </p>
                    </div>
                  </div>
                </li>
              </ul>
            </li>
            <li
              class="interactionsListItem"
            >
              <h3
                class="interactantTitle"
              >
                <a
                  href="/drugs/pancreatin/"
                >
                  Pancreatin
                </a>
              </h3>
              <ul
                class="messageList"
              >
                <li
                  class="message"
                >
                  <div
                    class=""
                  >
                    <div>
                      <span
                        class="substance-primary"
                        data-monograph-sid="_743802912"
                        data-sid="_743802912"
                      >
                        pancreatin
                      </span>
                       
                      <span
                        class="effectQualifier"
                      >
                        is predicted to
                      </span>
                       
                      <span
                        class="effect"
                      >
                        decrease
                      </span>
                       
                      <span
                        class="parameter"
                      >
                        the effects of
                      </span>
                       
                      <span
                        class="substance-primary"
                        data-monograph-sid="_426793441"
                        data-sid="_426793441"
                      >
                        acarbose
                      </span>
                      . 
                      <span
                        class="action"
                      >
                        Manufacturer advises avoid
                      </span>
                      .  
                    </div>
                    <dl
                      class="supplementaryInfo"
                    >
                       
                      <dt>
                        Evidence:
                      </dt>
                      <dd>
                        Theoretical
                      </dd>
                    </dl>
                  </div>
                </li>
              </ul>
            </li>
            <li
              class="interactionsListItem"
            >
              <h3
                class="interactantTitle"
              >
                <a
                  href="/drugs/testinteractant/"
                >
                  Test interactant
                </a>
              </h3>
              <ul
                class="messageList"
              >
                <li
                  class="message"
                >
                  <div
                    class=""
                  >
                    <div>
                      Test message 1
                    </div>
                  </div>
                </li>
                <li
                  class="message"
                >
                  <div
                    class="severeMessage"
                  >
                    <div>
                      Test message 2 (severe)
                    </div>
                    <dl
                      class="supplementaryInfo"
                    >
                      <dt>
                        <strong>
                          Severity:
                        </strong>
                      </dt>
                      <dd>
                        <strong>
                          Severe
                        </strong>
                      </dd>
                       
                    </dl>
                  </div>
                </li>
              </ul>
            </li>
          </ol>
        </section>
      </div>
    </div>
  </div>
</main>
`;

exports[`InteractantPage should match snapshot for graphql query 1`] = `
"
	query ($id: String) {
		bnfInteractant(id: { eq: $id }) {
			title
			drug {
				title
				slug
			}
			interactions {
				interactant {
					title
					drug {
						slug
					}
				}
				messages {
					additiveEffect
					evidence
					message
					severity
					severityOrder
				}
			}
		}
	}
"
`;
